Cabal/Distribution/FooBar.hs

```diff
@@ -0,0 +1,56 @@
+module Distribution.FooBar where
+
+import           Data.Maybe
+import qualified Data.Map as Map
+
+import           Distribution.Package
+import           Distribution.PackageDescription
+
+-- | Desugar a "build-tools" entry into proper a executable dependency if
+-- possible.
+--
+-- An entry can be so desguared in two cases:
+--
+-- 1. The name in build-tools matches a locally defined executable.  The
+--    executable dependency produced is on that exe in the current package.
+--
+-- 2. The name in build-tools matches a hard-coded set of known tools.  For now,
+--    the executable dependency produced is one an executable in a package of
+--    the same, but the hard-coding could just as well be per-key.
+--
+-- The first cases matches first.
+desugarBuildTool :: PackageDescription
+                 -> LegacyExeDependency
+                 -> Maybe ExeDependency
+desugarBuildTool pkg led =
+  if foundLocal
+  then Just $ ExeDependency (packageName pkg) toolName reqVer
+  else Map.lookup name whiteMap
+  where
+    LegacyExeDependency name reqVer = led
+    toolName = mkUnqualComponentName name
+    foundLocal = toolName `elem` map exeName (executables pkg)
+    whitelist = [ "hscolour", "haddock", "happy", "alex", "hsc2hs", "c2hs"
+                , "cpphs", "greencard"]
+    whiteMap  = Map.fromList $ flip map whitelist $ \n ->
+      (n, ExeDependency (mkPackageName n) (mkUnqualComponentName n) reqVer)
+
+-- | Get everything from "tool-depends", along with entries from "build-tools"
+-- that we know how to desugar.
+--
+-- This should almost always be used instead of just accessing the
+-- `toolDepends` field directly.
+getAllToolDependencies :: PackageDescription
+                       -> BuildInfo
+                       -> [ExeDependency]
+getAllToolDependencies pkg bi =
+  toolDepends bi ++ mapMaybe (desugarBuildTool pkg) (buildTools bi)
+
+-- | Does the given executable dependency map to this current package?
+--
+-- This is a tiny function, but used in a number of places. Note that the
+-- version bounds and components of the package are unchecked. This is because
+-- we sanitize exe deps so that the matching name implies these other
```

Comment:
I think the overall codebase doesn't use "sanitization" as the term here. Furthermore, not all checks in `PackageDescription.Check` are necessarily run all the time: a package can still be built even if some of the checks fail. It is better to say that there is some invariant that is expected to hold; in that case the invariant description should be at the data type definition.